File: jump_game2.py
# Time Complexity: O(n)
# Space Complexity: O(1)
class Solution:
    def jump(self, nums):
        count = 0
        current_jump = 0
        largest = 0
        #         We will check the the largest distance we can cover with  current jump
        #  If we reach to the end of current jump then we will make next jump
        for i in range(len(nums) - 1):
            largest = max(largest, i + nums[i])

            if i == current_jump:
                count += 1
                current_jump = largest

        return count

File: JumpGame-2.java

//Time complexity is O(N)
//Space complexity is O(1)
class Solution {
    public int jump(int[] nums) {
        int result=0;
        int max = 0;
        int curr = 0;
        for(int i=0;i<nums.length-1;i++){
            max =Math.max(max, i+nums[i]);
            if(curr==i){
                curr=max;
                result=result+1;
            }
        }
        return result;
    }
}